crates/recursor/src/recursor.rs

```diff
@@ -0,0 +1,454 @@
+// Copyright 2015-2022 Benjamin Fry <benjaminfry@me.com>
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+use std::{net::SocketAddr, time::Instant};
+
+use async_recursion::async_recursion;
+use futures_util::{future::select_all, FutureExt};
+
+use lru_cache::LruCache;
+use parking_lot::Mutex;
+use tracing::{debug, info, warn};
+
+use trust_dns_proto::{
+    op::Query,
+    rr::{RData, RecordType},
+};
+use trust_dns_resolver::{
+    config::{NameServerConfig, NameServerConfigGroup, Protocol, ResolverOpts},
+    dns_lru::{DnsLru, TtlConfig},
+    error::ResolveError,
+    lookup::Lookup,
+    name_server::NameServerPool,
+    Name, TokioConnectionProvider, TokioHandle,
+};
+
+use crate::{recursor_pool::RecursorPool, Error, ErrorKind};
+
+/// Set of nameservers by the zone name
+type NameServerCache = LruCache<Name, RecursorPool>;
```

Comment:
Style: I've kind of soured on type aliases; in my experience they make the code harder to understand (because of more opaque types) at not much benefit (no isolation of the inner type).